stringdtype/stringdtype/src/static_string.c

```diff
@@ -1,41 +1,76 @@
 #include "static_string.h"
 
-// allocates a new ss string of length len, filling with the contents of init
-ss *
-ssnewlen(const char *init, size_t len)
+int
+ssnewlen(const char *init, size_t len, ss *to_init)
 {
+    if ((to_init->buf != NULL) || (to_init->len != 0)) {
+        return -1;
+    }
+
     // one extra byte for null terminator
-    ss *ret = (ss *)malloc(sizeof(ss) + sizeof(char) * (len + 1));
+    char *ret_buf = (char *)malloc(sizeof(char) * (len + 1));
 
-    if (ret == NULL) {
-        return NULL;
+    if (ret_buf == NULL) {
+        return -1;
     }
 
-    ret->len = len;
+    to_init->len = len;
 
     if (len > 0) {
-        memcpy(ret->buf, init, len);
+        memcpy(ret_buf, init, len);
     }
 
-    ret->buf[len] = '\0';
+    ret_buf[len] = '\0';
+
+    to_init->buf = ret_buf;
+
+    return 0;
+}
 
-    return ret;
+void
+ssfree(ss *str)
+{
+    if (str->buf != NULL) {
+        free(str->buf);
+        str->buf = NULL;
+    }
+    str->len = 0;
 }
 
-// returns a new heap-allocated copy of input string *s*
-ss *
-ssdup(const ss *s)
+int
+ssdup(ss *in, ss *out)
 {
-    return ssnewlen(s->buf, s->len);
+    return ssnewlen(in->buf, in->len, out);
 }
 
-// returns a new, empty string of length len
-// does not do any initialization, the caller must
-// initialize and null-terminate the string
-ss *
-ssnewempty(size_t len)
+int
+ssnewemptylen(size_t num_bytes, ss *out)
 {
-    ss *ret = (ss *)malloc(sizeof(ss) + sizeof(char) * (len + 1));
-    ret->len = len;
-    return ret;
+    if (out->len != 0 || out->buf != NULL) {
+        return -1;
+    }
+
+    char *buf = (char *)malloc(sizeof(char) * (num_bytes + 1));
+
+    if (buf == NULL) {
+        return -1;
+    }
+
+    out->buf = buf;
+    out->len = num_bytes;
+
+    return 0;
+}
+
+static ss EMPTY = {0, "\0"};
+
+void
+load_string(char *data, ss **out)
+{
+    if (data == NULL) {
+        *out = &EMPTY;
+    }
+    else {
+        *out = (ss *)data;
+    }
```

Comment:
I don't think you should need this helper now?  data should never be NULL here? `data->buf` may be, but that is OK.

OTOH, I don't mind having the helper, it might be neat to experiment a bit more easily also.

Reply:
Yeah, my worry was what if code uses `data->buf`, even if `data->len` is zero. In practice that likely won't happen and if it does we should get a seg fault so it'll be really obvious.

Reply:
Ah, I guess since you are currently using/guaranteeing `\0` terminated strings it kind-of makes sense.

Reply:
But, you are not looking for `data == NULL` in that case, you would be looking for `data->len == 0` or `data->buf == NULL`.

Reply:
Argh, good call!